astropy/units/format/base.py

```diff
@@ -303,12 +288,11 @@ def _decompose_to_known_units(
             )
         if isinstance(unit, core.NamedUnit):
             try:
-                cls._get_unit_name(unit)
+                return cls._validate_unit(unit._get_format_name(cls.name))
             except ValueError:
                 if isinstance(unit, core.Unit):
                     return cls._decompose_to_known_units(unit._represents)
                 raise
-            return unit
```

Comment:
Moving the `return` to the `try`-block is valid if the string representation of the validated unit is the same as the string representation of the original unit. The validated units come from the `_units` dictionary that is associated with the parsing formatters. For the `CDS` formatter the values in `_units` are CDS units, not normal units, and for angstrom and solMass the string representations of the normal unit and the corresponding CDS unit were different, but that has been fixed in separate pull requests.